File: ABC/351-400/392/d.py
def main():
    n = int(input())
    ks = []
    arr = []
    ps = []
    for _ in range(n):
        ka = list(map(int,input().split()))
        k = ka[0]
        a = ka[1:]
        a = sorted(a)
        p = []
        for i in range(len(a)):
            if len(p) == 0:
                p.append((a[i],1))
            else:
                if p[-1][0] == a[i]:
                    aLast,cnt = p[-1]
                    p.pop(-1)
                    p.append((aLast,cnt+1))
                else:
                    p.append((a[i],1))

        ks.append(k)
        ps.append(p)
    ans = 0
    for i in range(n):
        for j in range(i+1,n):
            p1 = ps[i]
            p2 = ps[j]
            now = 0
            cur = 0
            for val,cnt in p2:
                while now < len(p1) and p1[now][0] < val:
                    now+=1
                if now < len(p1) and val == p1[now][0]:
                    #print(f"ok : {i},{j},{val}")
                    #print("calculated",cnt,len(),p1[now][1],len(p1))
                    cur += cnt/ks[i]*p1[now][1]/ks[j]
            ans = max(ans,cur)
    
    print(f"{ans}")

if __name__ == "__main__":
    main()

File: ABC/351-400/392/d.cpp
#include <bits/stdc++.h>
using namespace std;
#define rep(i,n) for(int i=0;i<n;++i)

int main(){
    int n;
    cin >> n;
    vector<int> k(n);
    vector<vector<int> > a(n);
    vector<map<int,double> > cnt(n);
    rep(i,n){
        cin >> k[i];
        a[i] = vector<int>(k[i]);
        rep(j,k[i]) cin >> a[i][j];
        rep(j,k[i]){
            cnt[i][a[i][j]]++;
        }
    }
    double ans = 0;
    rep(i,n){
        for(int j=i+1;j<n;j++){
            double now = 0;
            for(auto [val,c] : cnt[i]){
                if(!cnt[j].count(val)) continue;
                now += (double)cnt[i][val]/k[i]*cnt[j][val]/k[j];
            }
            ans = max(ans,now);
        }
    }
    printf("%.10f\n",ans);
    return 0;
}